numba/tests/test_boundscheck.py

```diff
@@ -0,0 +1,72 @@
+from __future__ import print_function, division, absolute_import
+
+import numpy as np
+
+from numba.compiler import compile_isolated, DEFAULT_FLAGS
+from numba import typeof
+from numba.types import float64
+from numba import unittest_support as unittest
+from .support import MemoryLeakMixin
+
+BOUNDSCHECK_FLAGS = DEFAULT_FLAGS.copy()
+BOUNDSCHECK_FLAGS.set('boundscheck', True)
+
+def basic_array_access(a):
+    return a[10]
+
+def slice_array_access(a):
+    # The first index (slice) is not bounds checked
+    return a[10:, 10]
+
+
+class TestBoundsCheck(MemoryLeakMixin, unittest.TestCase):
+    def test_basic_array_boundscheck(self):
+        a = np.arange(5)
+        # Check the numpy behavior to make sure the test is correct
+        with self.assertRaises(IndexError):
+            # TODO: When we raise the same error message as numpy, test that
+            # they are the same
+            basic_array_access(a)
+
+        at = typeof(a)
+        c_noboundscheck = compile_isolated(basic_array_access, [at],
+                                           flags=DEFAULT_FLAGS)
+        noboundscheck = c_noboundscheck.entry_point
+        c_boundscheck = compile_isolated(basic_array_access, [at],
+                                         flags=BOUNDSCHECK_FLAGS)
+        boundscheck = c_boundscheck.entry_point
+        # Check that the default flag doesn't raise
+        noboundscheck(a)
+        with self.assertRaises(IndexError):
+            boundscheck(a)
+
+    def test_slice_array_boundscheck(self):
+        a = np.ones((5, 5))
+        b = np.ones((5, 20))
+        with self.assertRaises(IndexError):
+            # TODO: When we raise the same error message as numpy, test that
+            # they are the same
+            slice_array_access(a)
+        # Out of bounds on a slice doesn't raise
+        slice_array_access(b)
+
+        at = typeof(a)
+        rt = float64[:]
+        c_noboundscheck = compile_isolated(slice_array_access, [at],
+                                           return_type=rt,
+                                           flags=DEFAULT_FLAGS)
+        noboundscheck = c_noboundscheck.entry_point
+        c_boundscheck = compile_isolated(slice_array_access, [at],
+                                         return_type=rt,
+                                         flags=BOUNDSCHECK_FLAGS)
+        boundscheck = c_boundscheck.entry_point
+        # Check that the default flag doesn't raise
+        noboundscheck(a)
+        noboundscheck(b)
+        with self.assertRaises(IndexError):
+            boundscheck(a)
```

Comment:
This line causes a memory leak error. I'm not clear why. Does it have to do with the fact that it raises an exception? The other test doesn't have an memory leak errors.

I was also having issues getting the `return_type` argument to `compile_isolated` correct for this test, so it could be related. 

Or it's possible there is a bug in the code. How can you debug memory leaks?

Reply:
If you are raising an exception from compiled code it will leak references (I'm assuming this is what you mean), it's a known issue. See point 6 of http://numba.pydata.org/numba-doc/latest/extending/overloading-guide.html#implementing-overload-for-numpy-functions typically tests which raise exceptions are split out from other tests so that the memory leak checks are only disabled for those tests which are known to leak.

Reply:
Ah that's good to hear. So no issues on my part. I was confused because the other test wasn't failing, so I thought it had something to to do with the different code path for tuple indices.

I'll split the test into the parts that raise and the parts that don't raise.

Reply:
Great, thanks. This is an approached used throughout the test base so will be consistent with that.